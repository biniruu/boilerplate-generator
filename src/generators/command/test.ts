import getCertainConditions from '@utils/certainConditions'
import type { SelectOptions } from '_types'

const getTestCommands = (configOptions: SelectOptions) => {
  const { hasCssModule, hasJest, hasNext, hasNuxt, hasReact, hasStorybook, hasTypescript, hasVue, hasWebpack } =
    getCertainConditions(configOptions)

  const testDevDependencies: string[] = []

  if (hasJest) {
    /**
     * @testing-library/dom (DOM Testing Library)
     * {@link https://testing-library.com/docs/dom-testing-library/intro/}
     *
     * @testing-library/jest-dom (jest-dom)
     * {@link https://github.com/testing-library/jest-dom#readme}
     *
     * @testing-library/user-event (user-event)
     * {@link https://github.com/testing-library/user-event#readme}
     *
     * jest (Jest)
     * {@link https://jestjs.io}
     *
     * jest-environment-jsdom
     * {@link https://github.com/jestjs/jest#readme}
     *
     * Jest Extended
     * {{@link https://jest-extended.jestcommunity.dev/docs}}
     *
     * jest-watch-typeahead
     * {@link https://github.com/jest-community/jest-watch-typeahead#jest-watch-typeahead}
     *
     * jsdom
     * {@link https://github.com/jsdom/jsdom#readme}
     *
     * msw (Mock Service Worker)
     * {@link https://mswjs.io}
     *
     * ts-jest
     * {@link https://kulshekhar.github.io/ts-jest/}
     * ts-jest is necessary when youâ€™re working in a development environment, since it provides types of jest configuration
     *
     *
     * If you have added 'jest', you don't need to add these packages: jest-resolve and jest-resolve-dependencies.
     * I recommend adding @types/jest even if it is not a TypeScript project because of its IntelliSense.
     */
    testDevDependencies.push(
      '@testing-library/dom',
      '@testing-library/jest-dom',
      '@testing-library/user-event',
      '@types/jest',
      'jest',
      'jest-environment-jsdom',
      'jest-extended',
      'jest-html-reporters',
      'jest-watch-typeahead',
      'jsdom',
      'msw',
      'ts-jest',
    )
    /**
     * identity-obj-proxy
     * {@link https://github.com/keyz/identity-obj-proxy#readme}
     */
    if (hasCssModule) {
      if (hasTypescript) {
        testDevDependencies.push('@types/identity-obj-proxy')
      }
      testDevDependencies.push('identity-obj-proxy')
    }
    /**
     * @testing-library/react (React Testing Library)
     * {@link https://testing-library.com/docs/react-testing-library/intro/})
     *
     * @testing-library/react-hooks (react-hooks-testing-library)
     * {@link https://github.com/testing-library/react-hooks-testing-library#readme}
     *
     * react-test-renderer (Test Renderer)
     * {@link https://legacy.reactjs.org/docs/test-renderer.html}
     */
    if (hasReact || hasNext) {
      if (hasStorybook) {
        testDevDependencies.push('@storybook/react', '@storybook/testing-library')
      }
      testDevDependencies.push(
        '@testing-library/react',
        '@testing-library/react-hooks',
        '@types/react-test-renderer',
        'react-test-renderer',
      )
    }
    /**
     * @testing-library/vue (Vue Testing Library)
     * {@link https://testing-library.com/docs/vue-testing-library/intro/}
     *
     * @vue/vue3-jest
     * {@link https://github.com/vuejs/vue-jest#readme}
     */
    if (hasNuxt || hasVue) {
      testDevDependencies.push('@testing-library/vue', '@vue/vue3-jest')
    }
    // TODO: Edit this jsdoc
    /**
     * @storybook/addon-a11y (Accessibility testing)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-actions (Action logger)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-essentials (Essential addons)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-interactions (Interact with components)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/addon-links (Link to other stories)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/builder-webpack5 (Webpack 5)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     *
     * @storybook/manager-webpack5 (Webpack 5 Storybook manager)
     * {@link XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX}
     */
    if (hasStorybook && !hasReact && !hasNext) {
      if (hasWebpack) {
        testDevDependencies.push('@storybook/builder-webpack5', '@storybook/manager-webpack5')
      }
      testDevDependencies.push(
        '@storybook/addon-a11y',
        '@storybook/addon-actions',
        '@storybook/addon-essentials',
        '@storybook/addon-interactions',
        '@storybook/addon-links',
      )
    }
  }

  return {
    testDevDependencies,
  }
}

export default getTestCommands
