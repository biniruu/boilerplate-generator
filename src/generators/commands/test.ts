import type { SelectOptions } from '_types'

const getTestCommands = (configOptions: SelectOptions) => {
  const hasJest = configOptions.jest
  const hasNext = configOptions.next
  const hasNuxt = configOptions.nuxt
  const hasPostcss = configOptions.postcss
  const hasReact = configOptions.react
  const hasTailwind = configOptions.tailwind
  const hasTypescript = configOptions.typescript
  const hasVue = configOptions.vue
  const hasCssModule = hasPostcss || hasTailwind

  const testDevDependencies: string[] = []

  if (hasJest) {
    if (hasCssModule) {
      if (hasTypescript) {
        testDevDependencies.push('@types/identity-obj-proxy')
      }
      testDevDependencies.push('identity-obj-proxy')
    }
    if (hasReact || hasNext) {
      testDevDependencies.push(
        '@testing-library/react',
        '@testing-library/react-hooks',
        '@types/react-test-renderer',
        'react-test-renderer',
      )
    }
    if (hasTypescript) {
      testDevDependencies.push('ts-jest')
    }
    if (hasNuxt || hasVue) {
      testDevDependencies.push('@testing-library/vue', '@vue/vue3-jest')
    }
    // If you have added 'jest', you don't need to add these packages: jest-resolve and jest-resolve-dependencies.
    // I recommend adding @types/jest even if it is not a TypeScript project because of its IntelliSense.
    testDevDependencies.push(
      '@testing-library/dom',
      '@testing-library/jest-dom',
      '@testing-library/user-event',
      '@types/jest',
      'jest',
      'jest-environment-jsdom',
      'jest-watch-typeahead',
      'jsdom',
      'msw',
    )
  }

  return {
    testDevDependencies,
  }
}

export default getTestCommands
