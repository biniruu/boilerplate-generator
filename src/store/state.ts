import { objOptions } from '@data/options';
import type { SelectOptions } from '_types';

import StateManager from './stateManager';

const initialState: SelectOptions = { ...objOptions };
const stateManager = new StateManager<SelectOptions>(initialState);

export default stateManager;
