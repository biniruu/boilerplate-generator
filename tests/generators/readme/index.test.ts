import generateReadme from '@generators/readme';
import { stateOptions } from '@store/state';

import { elemCode, options, setTextContent } from '../../setTextContent.test';

test('should return a README.md file', () => {
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a README.md file for Next.js project', () => {
  stateOptions.setState({
    ...options,
    next: true,
  });
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a README.md file for Next.js project with TypeScript', () => {
  stateOptions.setState({
    ...options,
    next: true,
    typescript: true,
  });
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a README.md file for React.js project', () => {
  stateOptions.setState({
    ...options,
    react: true,
  });
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a README.md file for Nuxt.js project', () => {
  stateOptions.setState({
    ...options,
    nuxt: true,
  });
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a README.md file for Webpack project', () => {
  stateOptions.setState({
    ...options,
    webpack: true,
  });
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a README.md file for Nuxt.js project with Webpack', () => {
  stateOptions.setState({
    ...options,
    nuxt: true,
    webpack: true,
  });
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});

test('should return a README.md file for Gatsby.js project', () => {
  stateOptions.setState({
    ...options,
    gatsby: true,
  });
  setTextContent({ content: generateReadme() });

  expect(elemCode?.textContent).toMatchSnapshot();
});
