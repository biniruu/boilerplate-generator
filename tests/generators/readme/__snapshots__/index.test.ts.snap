// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return a README.md file 1`] = `
"# How to set a development environment

1. Install the packages you selected
2. Add the created configuration files to the environment
3. Happy codingðŸ˜Ž"
`;

exports[`should return a README.md file for Gatsby.js project 1`] = `
"# How to set a development environment

1. Install the packages you selected
2. Add the created configuration files to the environment
3. Happy codingðŸ˜Ž

## Beginning Gatsby

If you want to use Gatsby with themes, use the starters provided by Gatsby. You can choose and install the necessary starter libraries from [Gatsby Starter Library](https://www.gatsbyjs.com/starters/).

## Creating Gatsby project

This boilerplate is configured based on the following options.

\`\`\`text
âœ“ Will you be using JavaScript or TypeScript?
â€¢ TypeScript

âœ“ Will you be using a CMS?
â€¢ No (or I'll add it later)

âœ“ Would you like to install a styling system?
â€¢ PostCSS

âœ“ Would you like to install additional features with other plugins?
â€¢ Add responsive images
â€¢ Add an automatic sitemap
â€¢ Generate a manifest file
â€¢ Add Markdown and MDX support
\`\`\`

If you don't want to configure the options yourself, use the command \`gatsby new [folder name]\`. Refer to the [Gatsby.js documentation](https://www.gatsbyjs.com/docs/tutorial/getting-started/part-1/#create-a-gatsby-site) to set up the basic settings for Gatsby."
`;

exports[`should return a README.md file for Next.js project 1`] = `
"# How to set a development environment

1. Create an app using \`yarn create next-app <my-app-name>\`
2. Install the packages you selected
3. Add the created configuration files to the environment
4. Happy codingðŸ˜Ž"
`;

exports[`should return a README.md file for Next.js project with TypeScript 1`] = `
"# How to set a development environment

1. Create an app using \`yarn create next-app <my-app-name> --typescript\`
2. Install the packages you selected
3. Add the created configuration files to the environment
4. Happy codingðŸ˜Ž"
`;

exports[`should return a README.md file for Nuxt.js project 1`] = `
"# How to set a development environment

1. Create an app using \`npx nuxi@latest init <project-name>\`
2. Install the packages you selected
3. Add the created configuration files to the environment
4. Happy codingðŸ˜Ž

## Setting PostCSS

You don't need to install the following packages since Nuxt.js already includes them ([See Nuxt.js documentation](https://nuxt.com/docs/getting-started/styling#using-postcss)):

- autoprefixer
- cssnano
- postcss-import
- postcss-url

Configure postcss in \`nuxt.config.ts\` instead of \`postcss.config.js\`.

## Modules

Building the project might take a long time when using modules from the Nuxt team and community. Unfortunately, I haven't found the solution. Refer to the Nuxt.js documentation for more information:

- [Modules](https://nuxt.com/modules)
- [Exploring Nuxt Modules](https://nuxt.com/docs/guide/concepts/modules#exploring-nuxt-modules)"
`;

exports[`should return a README.md file for Nuxt.js project with Webpack 1`] = `
"# How to set a development environment

1. Create an app using \`npx nuxi@latest init <project-name>\`
2. Install the packages you selected
3. Add the created configuration files to the environment
4. Happy codingðŸ˜Ž

## Initialising Webpack

This is not the best practise. Use your own methods.

1. Execute \`npx webpack init\`
2. If an error like \`For using this command you need to install: '@webpack-cli/generators' package.\` occurred, re-execute the command
3. Choose the options you need

## Setting PostCSS

You don't need to install the following packages since Nuxt.js already includes them ([See Nuxt.js documentation](https://nuxt.com/docs/getting-started/styling#using-postcss)):

- autoprefixer
- cssnano
- postcss-import
- postcss-url

Configure postcss in \`nuxt.config.ts\` instead of \`postcss.config.js\`.

## Modules

Building the project might take a long time when using modules from the Nuxt team and community. Unfortunately, I haven't found the solution. Refer to the Nuxt.js documentation for more information:

- [Modules](https://nuxt.com/modules)
- [Exploring Nuxt Modules](https://nuxt.com/docs/guide/concepts/modules#exploring-nuxt-modules)"
`;

exports[`should return a README.md file for React.js project 1`] = `
"# How to set a development environment

1. Create an app using \`yarn create vite <my-app-name> --template react-ts\`
2. Install the packages you selected
3. Add the created configuration files to the environment
4. Happy codingðŸ˜Ž"
`;

exports[`should return a README.md file for Webpack project 1`] = `
"# How to set a development environment

1. Install the packages you selected
2. Add the created configuration files to the environment
3. Happy codingðŸ˜Ž

## Initialising Webpack

This is not the best practise. Use your own methods.

1. Execute \`npx webpack init\`
2. If an error like \`For using this command you need to install: '@webpack-cli/generators' package.\` occurred, re-execute the command
3. Choose the options you need"
`;
